### DISCUSSION TITLE
Utils.merge unexpected merge result

### DISCUSSION THREAD

**Discussion by filenwind** [mikro-orm/mikro-orm#4101] [General]
[**Describe the bug**]

In fact, I don't know the behavior is intended or not, maybe is by design?

I found the em.assign with JSON column will make the data strange, 

So I dig source code found the result is merge by Mikro's Utils.merge

But it is different with other **_merge like function_** ex: lodash.merge, deepmerge

[**To Reproduce**]
[https://codesandbox.io/p/sandbox/silly-rain-23jce0?file=%2Findex.js](https://codesandbox.io/p/sandbox/silly-rain-23jce0?file=%2Findex.js)

You can see the log at right panel:

```
withNull

[mikro] { nestedData: null, moreDeep: { moreDeepData: null } }
[lodash] {
  nestedData: { foo: 'bar' },
  moreDeep: { moreDeepData: { foo: 'bar' } }
}
[deep] {
  nestedData: { foo: 'bar' },
  moreDeep: { moreDeepData: { foo: 'bar' } }
}

withString

[mikro] { nestedData: 'test', moreDeep: { moreDeepData: 'test' } }
[lodash] {
  nestedData: { foo: 'bar' },
  moreDeep: { moreDeepData: { foo: 'bar' } }
}
[deep] {
  nestedData: { foo: 'bar' },
  moreDeep: { moreDeepData: { foo: 'bar' } }
}

withObject

[mikro] {
  nestedData: { foo: 'bar' },
  moreDeep: { moreDeepData: { foo: 'bar' } }
}
[lodash] {
  nestedData: { foo: 'bar' },
  moreDeep: { moreDeepData: { foo: 'bar' } }
}
[deep] {
  nestedData: { foo: 'bar' },
  moreDeep: { moreDeepData: { foo: 'bar' } }
}
```

If origin data is not plain object, it won't be merge, source code here:

[https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/Utils.ts#L291](https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/Utils.ts#L291)

[**Expected behavior**]

Merge result same with lodash.merge or deepmerge

[**Additional context**]

Again, maybe this is by mikro's design, if yes, plz forgive me and close this

Thanks your help!

[**Versions**]

| Dependency | Version |
| - | - |
| mikro-orm | 5.6.13 |

**Comment by B4nan** [comment]
Indeed it is different, on purpose, the question is whether it should be different in this particular case. I'd consider that as an internal method, or implementation detail if you want. Would be good if you can provide a repro that shows what you want and what you get instead of comparing ORM internals with other libraries. I will have a closer look later today, on first sight, it looks like something to fix

The issue description sounds like you want to use custom classes in place of POJOs (although that is not part of the repro), for that you can extend the `PlainObject` class provided by the ORM so it understands what you want.

https://mikro-orm.io/docs/entity-helper#using-class-based-data

**Reply by B4nan** [reply]
Yeah, this indeed is a bug, the function was designed for config merging with defaults so it does not expect any structural differences.

**Reply by B4nan** [reply]
Let me know if this helps, available in `5.6.14-dev.4`

694ef281abf970434f42c7b620bfb81b1c794815

**Reply by filenwind** [reply]
Hi~ I tested, yes, it helps a lot!

But when I testing em.assign, It occur another bug, it's about **_compareObjects_**, 

I believe **_compareObjects_** is attempt to doing deep compare.

Here is the reproduce:

[https://codesandbox.io/p/sandbox/friendly-lovelace-bcvsxn?file=%2Findex.js](https://codesandbox.io/p/sandbox/friendly-lovelace-bcvsxn?file=%2Findex.js)

If object is create by **_Object.create(null)_**, it won't have **_valueOf_** or **_toString_**

I try to fix, and it works, maybe you can take a look for reference.
```javascript
if (a.valueOf && b.valueOf && a.valueOf !== Object.prototype.valueOf) {
    return a.valueOf() === b.valueOf();
}
if (a.toString && b.toString && a.toString !== Object.prototype.toString) {
    return a.toString() === b.toString();
}
```

Should I open issue for this?

Thank you very much for your help!!!!
